src/jsonschema.rs

```diff
@@ -68,17 +75,113 @@ struct Tag {
+        }
+    }
+
+    fn atom_into_ast(&self, data_type: Atom) -> ast::Tag {
+        match data_type {
+            Atom::Null => ast::Tag::new(ast::Type::Null, None, false),
+            Atom::Boolean => ast::Tag::new(ast::Type::Atom(ast::Atom::Boolean), None, false),
+            Atom::Number => ast::Tag::new(ast::Type::Atom(ast::Atom::Number), None, false),
+            Atom::Integer => ast::Tag::new(ast::Type::Atom(ast::Atom::Integer), None, false),
+            Atom::String => ast::Tag::new(ast::Type::Atom(ast::Atom::String), None, false),
+            Atom::Object => match &self.object.properties {
+                Some(properties) => {
+                    let mut fields: HashMap<String, Box<ast::Tag>> = HashMap::new();
+                    for (key, value) in properties {
+                        fields.insert(key.to_string(), Box::new(value.type_into_ast()));
+                    }
+                    ast::Tag::new(ast::Type::Object(ast::Object::new(fields)), None, false)
+                }
+                None => {
+                    // handle maps
+                    match (
+                        &self.object.additional_properties,
+                        &self.object.pattern_properties,
+                    ) {
+                        (Some(AdditionalProperties::Object(add)), Some(pat)) => {
+                            let value = ast::Tag::new(
+                                ast::Type::Union(ast::Union::new(vec![
+                                    add.type_into_ast(),
+                                    pat.type_into_ast(),
+                                ])),
+                                None,
+                                false,
+                            );
+
+                            ast::Tag::new(ast::Type::Map(ast::Map::new(None, value)), None, false)
+                        }
+                        (Some(AdditionalProperties::Object(tag)), None) | (None, Some(tag)) => {
+                            ast::Tag::new(
+                                ast::Type::Map(ast::Map::new(None, tag.type_into_ast())),
+                                None,
+                                false,
+                            )
+                        }
+                        _ => {
+                            // handle oneOf
+                            match &self.one_of {
+                                Some(vec) => {
+                                    let items =
+                                        vec.iter().map(|item| item.type_into_ast()).collect();
+                                    ast::Tag::new(
+                                        ast::Type::Union(ast::Union::new(items)),
+                                        None,
+                                        false,
+                                    )
+                                }
+                                None => {
+                                    ast::Tag::new(ast::Type::Atom(ast::Atom::JSON), None, false)
```

Comment:
This pattern matching arm is very deep, there may be a better way of writing this.